Moderated sessions UI message proposal (#10725)
gravitational/teleport: rfd/0043-kubeaccess-multiparty.md

state: implemented
---

# RFD 43 - Moderated Sessions

## What

Implement the ability to define RBAC policies with filters similar to those of [RFD 26](https://github.com/gravitational/teleport/blob/2fd6a88800604342bfa6277060b056d8bf0cbfb2/rfd/0026-custom-approval-conditions.md) that can be used to require others to be present in a session for it to be usable.

## Why

Heavily regulated and security-critical industries require that one or more participants with a certain role
are present in SSH and Kubernetes sessions and viewing it live to guarantee that
the operator does not perform any mistakes or acts of malice.

Such participants need to have the power to terminate a session immediately should anything go wrong.

To suit everyone this will need a more detailed configuration model based on rules
that can be used to define observers, their powers, and when and in what capacity they are required.

## Details

### Terminology

This RFD repeatedly refers to some specific nouns of which explanations can be found below:

> - Participant: Any user in a session regardless of mode, including the session initiator.

Reviewer: @xacrimon thoughts on adding definition for `initiator`?

Author: Done

> - Participant: Any user in a session regardless of mode, including the session initiator.

Reviewer: Nit:

```suggestion
- Participant: Any user in the session regardless of mode, including the session initiator.
```

Reviewer: @xacrimon what do you mean by "mode' here?

Author: Added a definition for that too.

> - Observer: A participant which only can view the session.

Reviewer: Nit:

```suggestion
- Observer: A participant who can only view the session.
```

> - Moderator: A participant who can view and terminate a session at any point.

Reviewer: Nit:

```suggestion
- Moderator: A participant who can view and terminate the session at any point.
```

Teleport >> MFA tap received.
```

> #### Broadcast messages

Reviewer: @xacrimon changes as discussed:

1. Define all 4 nouns (participant, observer, moderator, and peer) at the top
2. Change “launching session” to “connecting to $hostname using $protocol” for clarity
3. If moderator leaves (either voluntarily or failed MFA check), participants should see “Session terminated: participant requirements not met”
4. Provide a --participant-req flag that, if set, prints participant requirements after `Session paused, waiting for additional participants...`

for security and usability reasons.

Each broadcast message is prefixed with `Teleport > ` to indicate that this message
is injected by Teleport and not from the originating shell.

There are 9 kinds of broadcast messages:
- `Creating session with uuid <example-uuid>...`: Sent on session creation.
- `User <user> joined the session.`: Sent when a user joins the session.
- `User <user> left the session.`: Sent when a user leaves the session.
- `Lanching session...`: Sent when the session is launched and thus tranferred to a normal shell.
- `Session closed.`: Sent when the session is closed due to the termination of the shell.
- `Session forcefully terminated`: Sent when moderator forcefully terminates a session.
- `Session paused, waiting for additional participants...`: Sent when a session is paused due to lack of required participants.
- `Session resumed.`: Sent when a session has the required participants and is resumed.
- `Please tap your MFA key within 15 seconds.` and `Teleport >> MFA tap received.` are messages shown to moderators in sessions which are MFA-presence enabled.

##### Example

> This example illustrates how a group 3 users of which Alice is the initiator and Eve and Ben are two observers start a multiparty session. Below is a a series of events that happen that include what each user sees and actions taken.

Reviewer: Are _observers_ always moderators? I suggest being more clear here. For example:

```suggestion
This example illustrates how a TTY-enabled SSH that requires a moderator works. In this example, Alice is a participant and both Ben and Eve are moderators. Below is a series of events that happen and what each user sees as a result.
```
